src/extensions.jl

```diff
@@ -59,8 +59,13 @@ tags[:unionall] = d -> UnionAll(d[:var], d[:body])
 lower(x::Vector{Any}) = copy(x)
 lower(x::Vector{UInt8}) = x
 
-reinterpret_(::Type{T}, x) where T =
-  T[reinterpret(T, x)...]
+function reinterpret_(::Type{T}, x) where T
+  len = Int(length(x) * (sizeof(eltype(x)) / sizeof(T)))
+  GC.@preserve x begin
+    return unsafe_wrap(Array, Ptr{T}(pointer(x)), len)
```

Comment:
You may need to set `own = true` for `unsafe_wrap`.

Reply:
OK, that doesn't work on its own, but if I copy the memory then it works

```julia
function reinterpret_(::Type{T}, x::Vector{S}) where {T, S}
  length(x) < 1 && return T[]

  len = Int(length(x) * (sizeof(eltype(x)) / sizeof(T)))

  GC.@preserve x begin
    p = Ptr{UInt8}(pointer(x))
    t = Ptr{UInt8}(Base.Libc.calloc(len, sizeof(T)))
    ccall(:memset, Cvoid, (Ptr{UInt8}, Cint, Csize_t), t, 0, len * sizeof(T))
    ccall(:memcpy, Cvoid, (Ptr{UInt8}, Ptr{UInt8}, Csize_t), t, p, sizeof(x))
    Base.unsafe_wrap(Array, Ptr{T}(t), len; own=true)
  end
end
```

This brings the time down on my fork to 0.479740 seconds from 0.51184. It would be better to copy the bytes directly from the input stream and only memset the trailing bytes.

I also tried creating a second reference to the original memory and wrapping it, but `unsafe_wrap` just treats the Ref as a pointer AFAICT.

Reply:
If I add mmap and copy directly from the IOBuffers byte vector, then we get the following:

```
julia> @timev BSONqs.load("../../../julia/serbench/vgg19.bson"; mmap=true);
  0.250118 seconds (15.08 k allocations: 548.829 MiB, 19.26% gc time)
```